src/factotum_server/mod.rs

```diff
@@ -0,0 +1,242 @@
+            active: active_workers,
+        },
+        jobs: JobStatus {
+            max_queue_size: max_jobs,
+            in_queue: requests_queue.len(),
+            fail_count: 0,
+            success_count: 0
+        }
+    };
+    tx.send(result).expect("Server status channel receiver has been deallocated");
+}
+
+fn is_queue_full(query: Query<bool>, requests_queue: &mut VecDeque<JobRequest>, max_jobs: usize) {
+    let tx = query.status_tx;
+    let is_full = requests_queue.len() >= max_jobs;
+    tx.send(is_full).expect("Queue query channel receiver has been deallocated");
+}
+
+fn new_job_request<T: Persistence>(requests_channel: Sender<Dispatch>, requests_queue: &mut VecDeque<JobRequest>, primary_pool: ThreadPool, persistence: T, request: JobRequest) {
+    debug!("ADDING NEW JOB jobId:[{}]", request.job_id);
+    requests_queue.push_back(request.clone());
+    // Create entry in persistence storage
+    persist_entry(&persistence, request.job_id.clone(), request.clone(), JobState::Queued);
+    // Check queue size - return error if limit exceeded (not important right now)
+    if primary_pool.active_count() < primary_pool.max_count() {
+        requests_channel.send(Dispatch::ProcessRequest).expect("Job requests channel receiver has been deallocated");
+    } else {
+        info!("No threads available - waiting for a job to complete.")
+    }
+}
+
+fn process_job_request<T: 'static + Persistence + Send>(requests_channel: Sender<Dispatch>, requests_queue: &mut VecDeque<JobRequest>, primary_pool: ThreadPool, persistence: T, command_store: CommandStore) {
+    debug!("QUEUE SIZE = {}", requests_queue.len());
+    match requests_queue.pop_front() {
+        Some(request) => {
+            primary_pool.execute(move || {
+                debug!("PROCESSING JOB REQ jobId:[{}]", request.job_id);
+                // Update status in persistence storage
+                persist_entry(&persistence, request.job_id.clone(), request.clone(), JobState::Working);
+                let cmd_path = match command_store.get_command(::FACTOTUM) {
+                    Ok(path) => path,
+                    Err(e) => {
+                        error!("{}", e);
+                        requests_channel.send(Dispatch::RequestFailure(request)).expect("Job requests channel receiver has been deallocated");
+                        return
+                    }
+                };
+                let mut cmd_args = vec!["run".to_string(), request.factfile_path.clone()];
+                cmd_args.extend_from_slice(request.factfile_args.as_slice());
+                match command_store.execute(cmd_path, cmd_args) {
+                    Ok(output) => {
+                        trace!("{}", output);
+                        requests_channel.send(Dispatch::RequestComplete(request)).expect("Job requests channel receiver has been deallocated");
+                    },
+                    Err(e) => {
+                        error!("{}", e);
```

Comment:
Unable to return inside thread pool. Message is already being sent via channel. Will raise as an issue to look at in the future.